orderer/common/follower/follower_chain.go

```diff
@@ -286,146 +255,277 @@ func (c *Chain) run() {
 	}
 
+	err = c.pullAfterJoin()
+	if err != nil {
+		return errors.Wrap(err, "failed to pull after join block")
+	}
+
+	// Trigger creation of a new consensus.Chain.
+	c.logger.Info("Block pulling finished successfully, invoking chain creation callback")
+	c.halt()
+	c.chainCreationCallback()
+
 	return nil
 }
 
 // pullUpToJoin pulls blocks up to the join-block height without inspecting membership on fetched config blocks.
 // It checks whether the chain was stopped between blocks.
 func (c *Chain) pullUpToJoin() error {
 	targetHeight := c.joinBlock.Header.Number + 1
-	firstBlockToPull := c.ledgerResources.Height()
-	c.logger.Debugf("first block to pull: %d, target height: %d", firstBlockToPull, targetHeight)
-	if firstBlockToPull >= targetHeight {
-		c.logger.Infof("Target height (%d) is <= to our ledger height (%d), skipping pulling", targetHeight, firstBlockToPull)
+	if c.ledgerResources.Height() >= targetHeight {
+		c.logger.Infof("Target height (%d) is <= to our ledger height (%d), skipping pulling", targetHeight, c.ledgerResources.Height())
 		return nil
 	}
 
-	puller, err := c.createBlockPullerFunc()
+	var err error
+	c.blockPuller, err = c.createBlockPullerFunc() // TODO replace with a factory that has a join block as input
+	if err != nil {                                //This should never happen since we check the join-block before we start.
+		return errors.Wrapf(err, "error creating block puller")
+	}
+	defer c.blockPuller.Close()
+
+	err = c.pullUntilLatestWithRetry(targetHeight, false)
+	if err != nil {
+		return err
+	}
+
+	c.logger.Infof("Pulled blocks from %d until %d", c.firstHeight, targetHeight-1)
+	return nil
+}
+
+// pullAfterJoin pulls blocks continuously, inspecting the fetched config blocks for membership.
+// On every config block, it renews the BlockPuller, to take in the new configuration.
+// It will exit with 'nil' if it detects a config block that indicates the orderer is a member of the cluster.
+// It checks whether the chain was stopped between blocks.
+func (c *Chain) pullAfterJoin() error {
+	var err error
+
+	c.blockPuller, err = c.createBlockPullerFunc() // TODO replace with a factory that has last config block as input
 	if err != nil {
-		c.logger.Errorf("Error creating block puller: %s", err)
 		return errors.Wrap(err, "error creating block puller")
 	}
-	defer puller.Close()
+	defer c.blockPuller.Close()
+
+	heightPollInterval := c.options.HeightPollMinInterval
+	for {
+		// Check membership
+		isMember, errMem := c.clusterConsenter.IsChannelMember(c.lastConfig)
+		if errMem != nil {
+			return errors.Wrap(err, "failed to determine channel membership from last config")
+		}
+		if isMember {
+			c.setStatusReport(types.ClusterRelationMember, types.StatusActive)
+			return nil
+		}
+
+		// Poll for latest network height to advance beyond ledger height.
+		var latestNetworkHeight uint64
+	heightPollLoop:
+		for {
+			endpoint, networkHeight, errHeight := cluster.LatestHeightAndEndpoint(c.blockPuller)
```

Comment:
So I suppose this is why we are creating the block puller externally, instead of inside of the `pullUntilLatest` with retry.  But I'm not convinced that we need this call here.

Why not have `pullUntilLatest` simply pull until the latest, in both paths?  As I've mentioned in other places in the review, I absolutely don't see why it's critical that we stop pulling after we get to the 'Join Block'.  We should replicate the chain up to its highest height, and then decide if we are members.

Then essentially, in the 'pullUntilJoin' path, you would create a block puller from the join block, and let it go.  In the 'pullAfterJoin' path, you would look at the latest config, create the block puller, and let it go.

Reply:
Both paths call `pullUntilLatestWithRetry` only. `pullUntilLatestWithRetry` calls `pullUntilLatest`. I could have consolidated both these function in to one, but that would have created deed nesting and a long function, longer than my screen can show. 

The two paths require slightly different block pullers, see full comment below.

Reply:
Yes, sloppy commenting on my part "`pullUntilLatest` with retry" should have been `pullUntilLatestWithRetry`.  No qualms with splitting that particular function into two pieces.  I was simply suggesting that there was no need to parameterize the function, that we could have it simply query the latest height, and then pull until this height.